diff --git a/node_modules/hyperdht/index.js b/node_modules/hyperdht/index.js
index ea8472c..25df899 100644
--- a/node_modules/hyperdht/index.js
+++ b/node_modules/hyperdht/index.js
@@ -16,7 +16,7 @@ const ConnectionPool = require('./lib/connection-pool')
 const { STREAM_NOT_CONNECTED } = require('./lib/errors')
 
 class HyperDHT extends DHT {
-  constructor (opts = {}) {
+  constructor (opts, sig, dht_keys, keychain = {}) {
     const port = opts.port || 49737
     const bootstrap = opts.bootstrap || BOOTSTRAP_NODES
 
@@ -24,7 +24,12 @@ class HyperDHT extends DHT {
 
     const { router, persistent } = defaultCacheOpts(opts)
 
-    this.defaultKeyPair = opts.keyPair || createKeyPair(opts.seed)
+    this.checkedSigs = [sig]
+    this.sig = sig
+    this.keychain = keychain
+    this.dht_keys = dht_keys
+    this.defaultKeyPair = dht_keys || createKeyPair(opts.seed)
+    
     this.listening = new Set()
 
     this._router = new Router(this, router)
@@ -50,8 +55,8 @@ class HyperDHT extends DHT {
     })
   }
 
-  connect (remotePublicKey, opts) {
-    return connect(this, remotePublicKey, opts)
+  connect (remotePublicKey, opts ) {
+    return connect(this, remotePublicKey, opts, this.sig)
   }
 
   createServer (opts, onconnection) {
diff --git a/node_modules/hyperdht/lib/connect.js b/node_modules/hyperdht/lib/connect.js
index de9eef1..e8c883a 100644
--- a/node_modules/hyperdht/lib/connect.js
+++ b/node_modules/hyperdht/lib/connect.js
@@ -29,7 +29,7 @@ const {
   SUSPENDED
 } = require('./errors')
 
-module.exports = function connect (dht, publicKey, opts = {}) {
+module.exports = function connect (dht, publicKey, opts, sig = {}) {
   const pool = opts.pool || null
 
   if (pool && pool.has(publicKey)) return pool.get(publicKey)
@@ -69,6 +69,7 @@ module.exports = function connect (dht, publicKey, opts = {}) {
     onsocket: null,
     sleeper: new Sleeper(),
     encryptedSocket,
+    sig,
 
     // Relay state
     relayTimeout: null,
@@ -364,7 +365,8 @@ async function connectThroughNode (c, address, socket) {
       secretStream: {},
       relayThrough: c.relayThrough
         ? { publicKey: c.relayThrough, token: c.relayToken }
-        : null
+        : null,
+      mid: c.sig
     })
     if (isDone(c)) return
   }
@@ -390,6 +392,23 @@ async function connectThroughNode (c, address, socket) {
   
   const hs = c.handshake.final()
 
+  //firewall function does not send payload so we must check here for now
+  //Checking if the remotePublickey is signed with our common privatekey
+  let error = false
+  if (!payload.mid) error = true
+  
+  const checked = checkSignature(payload.mid, hs.remotePublicKey, c)
+  //Check if this id is already used by another active connection
+  if (c.dht.checkedSigs.some(a => payload.mid === a)) error = true
+  if (!checked) error = true
+
+  c.dht.checkedSigs.push(payload.mid)
+  
+  if (error || !checked) {
+      maybeDestroyEncryptedSocket(c, SERVER_ERROR('Verification failed'))
+      return
+  }
+
   c.handshake = null
   c.request = null
   c.requesting = false
@@ -452,6 +471,12 @@ async function connectThroughNode (c, address, socket) {
   c.session.destroy()
 }
 
+const checkSignature = (signature, remotePublicKey, c) => {
+  const keys = c.dht.keychain.get()
+  const verify = keys.verify(remotePublicKey, signature, keys.publicKey)
+  return verify ? true : false
+ }
+
 async function updateHolepunch (c, peerAddress, relayAddr, payload) {
   const holepunch = await c.dht._router.peerHolepunch(c.target, {
     id: c.connect.payload.holepunch.id,
diff --git a/node_modules/hyperdht/lib/messages.js b/node_modules/hyperdht/lib/messages.js
index 86d83ee..c1b0ea2 100644
--- a/node_modules/hyperdht/lib/messages.js
+++ b/node_modules/hyperdht/lib/messages.js
@@ -155,7 +155,7 @@ const relayThroughInfo = {
 
 exports.noisePayload = {
   preencode (state, m) {
-    state.end += 4 // version + flags + error + firewall
+    state.end += 68 // version + flags + error + firewall + newpub + id
     if (m.holepunch) holepunchInfo.preencode(state, m.holepunch)
     if (m.addresses4 && m.addresses4.length) ipv4Array.preencode(state, m.addresses4)
     if (m.addresses6 && m.addresses6.length) ipv6Array.preencode(state, m.addresses6)
@@ -184,6 +184,7 @@ exports.noisePayload = {
     if (m.udx) udxInfo.encode(state, m.udx)
     if (m.secretStream) secretStreamInfo.encode(state, m.secretStream)
     if (m.relayThrough) relayThroughInfo.encode(state, m.relayThrough)
+    if (m.mid) c.fixed64.encode(state, m.mid)
   },
   decode (state) {
     const version = c.uint.decode(state)
@@ -200,7 +201,8 @@ exports.noisePayload = {
         addresses6: [],
         udx: null,
         secretStream: null,
-        relayThrough: null
+        relayThrough: null,
+        mid: null
       }
     }
 
@@ -215,7 +217,8 @@ exports.noisePayload = {
       addresses6: (flags & 4) !== 0 ? ipv6Array.decode(state) : [],
       udx: (flags & 8) !== 0 ? udxInfo.decode(state) : null,
       secretStream: (flags & 16) !== 0 ? secretStreamInfo.decode(state) : null,
-      relayThrough: (flags & 32) !== 0 ? relayThroughInfo.decode(state) : null
+      relayThrough: (flags & 32) !== 0 ? relayThroughInfo.decode(state) : null,
+      mid: c.fixed64.decode(state)
     }
   }
 }
diff --git a/node_modules/hyperdht/lib/server.js b/node_modules/hyperdht/lib/server.js
index ca19d4a..b3e710b 100644
--- a/node_modules/hyperdht/lib/server.js
+++ b/node_modules/hyperdht/lib/server.js
@@ -22,7 +22,8 @@ module.exports = class Server extends EventEmitter {
 
     this.dht = dht
     this.target = null
-
+    this.sig = dht.sig
+    this.keychain = dht.keychain
     this.closed = false
     this.firewall = opts.firewall || (() => false)
     this.holepunch = opts.holepunch || (() => true)
@@ -337,7 +338,8 @@ module.exports = class Server extends EventEmitter {
         secretStream: {},
         relayThrough: relayThrough
           ? { publicKey: relayThrough, token: hs.relayToken }
-          : null
+          : null,
+        mid: this.sig
       })
     } catch (err) {
       safetyCatch(err)
