diff --git a/node_modules/hyperbeam/index.js b/node_modules/hyperbeam/index.js
index 438cfb1..d2a3966 100644
--- a/node_modules/hyperbeam/index.js
+++ b/node_modules/hyperbeam/index.js
@@ -1,9 +1,9 @@
 const { Duplex } = require('streamx')
-const sodium = require('sodium-universal')
 const b4a = require('b4a')
 const queueTick = require('queue-tick')
 const b32 = require('hi-base32')
 const DHT = require('@hyperswarm/dht')
+const nacl = require('tweetnacl')
 
 module.exports = class Hyperbeam extends Duplex {
   constructor (key, options) {
@@ -186,7 +186,6 @@ function fromBase32 (str) {
 }
 
 function randomBytes (length) {
-  const buffer = b4a.alloc(length)
-  sodium.randombytes_buf(buffer)
-  return buffer
+  let random = Buffer.alloc(32).fill(nacl.randomBytes(32).toString('hex'))
+  return random
 }
